### ISSUE TITLE
Performance tests: deploy simulated agents massively

### ISSUE THREAD

**Issue by BraulioV** [wazuh/wazuh-qa#1240]
The deployment of the simulated agents for the performance tests of the cluster doesn't scale as expected. For 1K agents, the test works fine but when the number of the agents grows up to 25K, the pipeline can't handle them for several reasons:
- The deployment of the ECS tasks takes too much time
- It consumes too many resources in the Jenkins node.
- The number of simulated agents per instance is not optimized. In this case, we only need to simulate agents that will only send the keep-alive message to the manager and receive it. So we can increase the number of agents per host.

## Tasks
- [x] **1**: Restrict the type instance for this use case to c5.xlarge.
- [x] **2**: Improve the deployment of the agent_simulator tasks. AWS allows the deployment of up to 10 tasks at the same time from the same definition. 
- [x] **3**:  Create a new AMI definition in the Jenkins cloud to give it a specific label. Necessary to launch a build per instance and make sure that the host has enough resources.
- [x] **4**: Deploy the ECS tasks in parallel. 
- [x] **5**: Optimize the number of agents that can be deployed in a c5.xlarge instance. 
- [x] **6**: Deploy up to 50K agents.
- [x] **7**: API log artifacts.

**Comment by jmv74211** [comment]
## Task 1

This task has been done in this PR https://github.com/wazuh/wazuh-jenkins/pull/2498

The logic of the instances deployment is as follows:

- If the testing time exceeds the threshold (45 min), `c5xlarge` is selected for both agent and manager.

- `auto` mode will always choose `t2.medium` unless the test time does not exceed the threshold, or the user has not explicitly selected `c5xlarge`.

- If the user chooses `c5xlarge`, it will be selected

- If the test mode is `AGENTS`, then the type of instances used to deploy agents will be `c5xlarge`.

**Comment by jmv74211** [comment]
# Task  2

I have added a new condition in the cluster pipeline so that when running in agent mode, the ECSTasks of the agents are grouped 10 by 10 (maximum set by AWS), so that 10 ECSTasks are deployed in the same request.

The changes mentioned are the following https://github.com/wazuh/wazuh-jenkins/commit/0794e4e9df93b61b8d1f5251a7f51cd80132cabc and https://github.com/wazuh/wazuh-jenkins/commit/1a80832d04cf84e8952e9bb51fdf8a1770cca93a

**Comment by jmv74211** [comment]
# Task 3

From now on, each cluster performance test build will be launched on a single independent Jenkins node. The reason is to avoid saturating the host where the build is being executed because when deploying in parallel a number of N instances, the host can be temporarily saturated with a single build.

This configuration has been applied in both staging and production workspaces.

For this purpose, a new AMI with an executor has been added to the list and assigned the label `cluster_test`

The result would be as follows (A separate instance for each build):

![result](https://user-images.githubusercontent.com/23462183/115877383-aa327980-a447-11eb-852a-1d3833d6a4bb.png)

**Comment by jmv74211** [comment]
Also during one of the tests, a bug was discovered in the allocation of agents to ECSTasks. I have proceeded to make the necessary fix.

The changes made are the ones indicated in this PR https://github.com/wazuh/wazuh-jenkins/pull/2503

**Comment by jmv74211** [comment]
## Task 5 

I have been running several tests, and I have concluded that for the instance type `c5.xlarge` (4vCPUs and 8GB RAM) a total of **800 agents** can be deployed by sending only the `keep-alives` (using 90% of host memory).

This has been tested on an EC2 instance, so we have to see if an ECSTask with the same resources has the same results.

**Comment by jmv74211** [comment]
## Task 4 

I have been investigating the best way to achieve parallel deployment in the shortest possible time. To do this, in AGENTS mode I have been grouping the ECSTasks in batches of 10 (maximum allowed by AWS) and launching these requests in parallel in Jenkins.

Given that for now, the upper limit of agents to launch is 100k and each ECSTask can simulate 800 agents sending keep-alives messages, the maximum number of ECSTasks is 125 which grouped in sizes of 10, gives a total of 13 groups.

I then tried launching all this in parallel from Jenkins, but it didn't work, as there seems to be a time limit on AWS requests.

```
An error occurred (ThrottlingException) when calling the RunTask operation (reached max retries: 2): Rate exceeded
```

To solve this, I have added a delay per group of about 15 seconds between each one for each parallel request.

```
group 1 --> t(0)
group 2 --> t(15)
group 3 --> t(30)
...
```

Also, I have been calculating the time it takes to execute each group of requests to AWS, with the following results:

```
Task 0 started at Mon Apr 26 10:25:38 UTC 2021 and stoped at Mon Apr 26 10:26:50 UTC 2021,  --> 1m 12s
Task 1 started at Mon Apr 26 10:25:54 UTC 2021 and stoped at Mon Apr 26 10:27:00 UTC 2021,  --> 1m 6s
Task 2 started at Mon Apr 26 10:26:10 UTC 2021 and stoped at Mon Apr 26 10:27:16 UTC 2021,  --> 1m 6s
Task 3 started at Mon Apr 26 10:26:25 UTC 2021 and stoped at Mon Apr 26 10:27:44 UTC 2021,  --> 1m 19s
Task 4 started at Mon Apr 26 10:26:40 UTC 2021 and stoped at Mon Apr 26 10:27:58 UTC 2021,  --> 1m 18s
Task 5 started at Mon Apr 26 10:26:55 UTC 2021 and stoped at Mon Apr 26 10:28:09 UTC 2021,  --> 1m 14s
Task 6 started at Mon Apr 26 10:27:10 UTC 2021 and stoped at Mon Apr 26 10:28:33 UTC 2021,  --> 1m 23s
Task 7 started at Mon Apr 26 10:27:25 UTC 2021 and stoped at Mon Apr 26 10:29:10 UTC 2021,  --> 1m 45s
Task 8 started at Mon Apr 26 10:27:40 UTC 2021 and stoped at Mon Apr 26 10:28:54 UTC 2021,  --> 1m 14s
Task 9 started at Mon Apr 26 10:27:55 UTC 2021 and stoped at Mon Apr 26 10:29:04 UTC 2021,  --> 1m 9s
Task 10 started at Mon Apr 26 10:28:10 UTC 2021 and stoped at Mon Apr 26 10:29:23 UTC 2021, --> 1m 13s
Task 11 started at Mon Apr 26 10:28:25 UTC 2021 and stoped at Mon Apr 26 10:29:39 UTC 2021, --> 1m 14s
Task 12 started at Mon Apr 26 10:28:40 UTC 2021 and stoped at Mon Apr 26 10:29:54 UTC 2021  --> 1m 14s
```

Thus, we know that at most it can take around 2 minutes to start each group.

With this data, we can calculate the total time it can take to deploy all these ECSTasks.

```
deployment_time = 2 + ((groups_number * 15) / 60) 
```

So deployment for 100k agents can take around 5m 15s. This time will be added to the monitoring time so that everything expected is monitored.

To sum up:

- **Total number of ECSTasks for 100k agents**: `125`
- **Time taken to deploy 100k agents**:  `5m 15s`

**Comment by jmv74211** [comment]
## Task 6

Having finished everything related to the parallel deployment, I have been testing whether everything works as expected by deploying a massive number of agents.

I started from the amount of 50k.

At the beginning I saw that around 15k agents successfully connected. Researching I found the possible reasons:

- On the master node side, authd error errors were being displayed because it was receiving too many connections simultaneously, so it was not registering the agents and connecting correctly

   ```
   2021/04/26 13:33:23 wazuh-authd: ERROR: Too many connections. Rejecting.
   ```

- On the workers side, I also observed the following problem:

   ```
   2021/04/26 13:40:16 wazuh-remoted: ERROR: TCP peer [352] at 172.31.4.72: Bad file descriptor (9).
   ```

It seems that the machine was running out of file descriptors and connections.

After observing this, the following measures were taken:

- I have added a timeout between the deployment of ECSTasks groups (https://github.com/wazuh/wazuh-jenkins/commit/5402a19efda9dbaf22f37c13537a2571e0c686bb).

- On wazuh cluster nodes, I have changed the maximum number of requests queued to a listen socket with `sysctl` (https://github.com/wazuh/wazuh-jenkins/commit/1df7b8a80c79ccd9e40224bedab0347ed21284a2).

- The ulimits of the ECSTasks to be deployed by the agents have been modified to the following values:

    ```
    nofile	Soft limit 262144 Hard limit 262144
    ```

- A number of retries have been added when sending the registration request to the manager by the agents (https://github.com/wazuh/wazuh-qa/pull/1253).

Following this, we have seen significant improvements, for example in a build of 50k agents, we managed to register and connect 48093.

We are still investigating how to improve this figure so that there is no problem related to this issue whatever the number of agents to be deployed.

**Comment by jmv74211** [comment]
## Task 7

At the request of the framework team, each of the `api.log` of the instances that make up the wazuh cluster has been collected and added to the artifacts generated in the Jenkins build.

All related changes are in https://github.com/wazuh/wazuh-jenkins/commit/a8ca01aead60a8a43e09ed0ed07f1f21846d3894

**Comment by BraulioV** [comment]
Closed by https://github.com/wazuh/wazuh-jenkins/pull/2518
